File: php/planet-time/src/InterplanetTime/Scheduling.php
<?php
declare(strict_types=1);

namespace InterplanetTime;

/**
 * Meeting window finder — ported from planet-time.js findMeetingWindows().
 */
final class Scheduling
{
    private function __construct() {}

    /**
     * Find overlapping work windows for two planets over a range of Earth days.
     *
     * @param string $planetA  Planet key (e.g. 'earth', 'mars')
     * @param string $planetB  Planet key
     * @param int    $fromMs   Start UTC ms
     * @param int    $earthDays Number of Earth days to scan (default 7)
     * @return MeetingWindow[]
     */
    public static function findMeetingWindows(
        string $planetA,
        string $planetB,
        int    $fromMs,
        int    $earthDays = 7,
    ): array {
        $STEP_MS = 15 * 60 * 1000; // 15-minute step
        $endMs   = $fromMs + $earthDays * Constants::EARTH_DAY_MS;

        $windows     = [];
        $inWindow    = false;
        $windowStart = 0;

        for ($t = $fromMs; $t < $endMs; $t += $STEP_MS) {
            $ta = Time::getPlanetTime($planetA, $t);
            $tb = Time::getPlanetTime($planetB, $t);
            $overlap = $ta->isWorkHour && $tb->isWorkHour;

            if ($overlap && !$inWindow) {
                $inWindow    = true;
                $windowStart = $t;
            } elseif (!$overlap && $inWindow) {
                $inWindow = false;
                $dur      = (int)(($t - $windowStart) / 60000);
                $windows[] = new MeetingWindow($windowStart, $t, $dur);
            }
        }

        if ($inWindow) {
            $dur       = (int)(($endMs - $windowStart) / 60000);
            $windows[] = new MeetingWindow($windowStart, $endMs, $dur);
        }

        return $windows;
    }
}

File: kotlin/planet-time/src/main/kotlin/interplanet/time/Scheduling.kt
package interplanet.time

/**
 * Scheduling.kt — findMeetingWindows
 * Ported from planet-time.js v1.1.0
 */

/**
 * Find overlapping work windows between two planets over N Earth days.
 * Step size: 15 minutes.
 */
fun findMeetingWindows(
    planetA: Planet,
    planetB: Planet,
    earthDays: Int = 7,
    startMs: Long = System.currentTimeMillis()
): List<MeetingWindow> {
    val stepMs = 15 * 60000L
    val endMs = startMs + earthDays * EARTH_DAY_MS
    val windows = mutableListOf<MeetingWindow>()
    var inWindow = false
    var windowStart = 0L

    var t = startMs
    while (t < endMs) {
        val ta = getPlanetTime(planetA, t)
        val tb = getPlanetTime(planetB, t)
        val overlap = ta.isWorkHour && tb.isWorkHour
        if (overlap && !inWindow) {
            inWindow = true
            windowStart = t
        }
        if (!overlap && inWindow) {
            inWindow = false
            windows.add(MeetingWindow(
                startMs = windowStart,
                endMs = t,
                durationMin = ((t - windowStart) / 60000L).toInt()
            ))
        }
        t += stepMs
    }
    if (inWindow) {
        windows.add(MeetingWindow(
            startMs = windowStart,
            endMs = endMs,
            durationMin = ((endMs - windowStart) / 60000L).toInt()
        ))
    }
    return windows
}
